File: 2017/day19/app/coordinate.ts
export class Coordinate {
    constructor(row: number, col: number) {
        this.row = row;
        this.col = col;
    }

    row: number;
    col: number;

    up(): Coordinate {
        return new Coordinate(this.row - 1, this.col);
    }

    down(): Coordinate {
        return new Coordinate(this.row + 1, this.col);
    }

    left(): Coordinate {
        return new Coordinate(this.row, this.col-1);
    }

    right(): Coordinate {
        return new Coordinate(this.row, this.col + 1);
    }

    move(direction: string): Coordinate {
        var map: { [direction: string]: () => Coordinate} = {
            "up": this.up,
            "down": this.down,
            "left": this.left,
            "right": this.right
        };

        return map[direction]();
    }

    toString(): string {
        return `(${this.row}, ${this.col})`;
    }
}

File: java/src/main/java/org/dathan/aoc/Coordinate.java
package org.dathan.aoc;

public record Coordinate(int x, int y) {
    public static final Coordinate ORIGIN = new Coordinate(0, 0);

    public Coordinate left() {
        return new Coordinate(x - 1, y);
    }

    public Coordinate right() {
        return new Coordinate(x + 1, y);
    }

    public Coordinate up() {
        return new Coordinate(x, y + 1);
    }

    public Coordinate down() {
        return new Coordinate(x, y - 1);
    }

    public Coordinate move(char direction) {
        return switch (direction) {
            case '<' -> left();
            case '>' -> right();
            case '^' -> up();
            case 'V', 'v' -> down();
            default -> throw new RuntimeException("Unrecognized direction " + direction);
        };
    }
}
